### DISCUSSION TITLE
Require multiline defs to be enclosed by `{...}`

### DISCUSSION THREAD

**Discussion by JonathanStarup** [flix/flix#5490] [Ideas]
This seems required for proper inner function syntax based on discussions with @mlutze.

This would allow function like
```
def plusOne(x: Int32): Int32 = x + 1

def addSquares(x: Int32, y: Int32): Int32 = {
    let xSquare = x * x;
    let ySquare = y * y;
    sSquare = ySquare
}
```
And disallow
```
def addSquares(x: Int32, y: Int32): Int32 =
    let xSquare = x * x;
    let ySquare = y * y;
    sSquare = ySquare
```

One way to implement this would be that the def expression cannot have let without {}, that would disallow
```
def test(): Int32 = let x = 42; x +3
```
Another way would be to only parse the remaining line if the expression does not begin with { in which case the above program would be allowed.

This could also be considered for lambdas, much like scala, to force
```
x ->
    let x = 42;
    let y = 12;
    x + y
```
to become
```
x -> {
    let x = 42;
    let y = 12;
    x + y
}
```

Personally I like the rule that lets\defs require a block. This is a simple rule to remember and is less strict on the syntax. This should also be done for statements in a natual way. Simply said: there is only `;` inside of a block. This also allows multiline oneliners like
```
def complex(x: SimpleData): CoolData = x |>
    CoolData.verify |>
    CoolData.transfrom |>
    CoolData.finalize
```

This problems also shows in nested defs where the def body is an expression,which would allow let-binding, but this can never lead to a parsing program because the parser doesn't know when the ; ends the inner def or is just a stm in the inner def.
```
def outer(): Int32 =
    def inner() =
        let local = 42;
        println(local);
    inner()
```
This should "technically" be a valid program but is a parser error

The result of this is that the parser has a context on which to know whether a def is nested (inside of {}) or a new toplevel def. This means that they can be identical and still be parsed correctly.

**Comment by magnus-madsen** [comment]
Do you have some examples of what you want to alive/disallow?

**Comment by JonathanStarup** [comment]
> Do you have some examples of what you want to alive/disallow?

I've updated the description.

**Comment by magnus-madsen** [comment]
OK, I understand what you want, but I am not sure I understand what problem it solves. What is wrong with:

```
def addSquares(x: Int32, y: Int32): Int32 =
    let xSquare = x * x;
    let ySquare = y * y;
    sSquare = ySquare
```

It seems very regular to me.

**Comment by JonathanStarup** [comment]
I just think it leads to some weird errors, parsing difficulties, and this kind of parsing change is needed at least in letrec. Lets can probably be left alone on non letrecs but I think its good style to enforce and nice for consistency.
